components/builder-depot/src/server.rs

```diff
@@ -982,23 +1079,61 @@ fn render_package(pkg: &depotsrv::Package, should_cache: bool) -> IronResult<Res
 }
 
 fn promote_package(req: &mut Request) -> IronResult<Response> {
-    let depot = req.get::<persistent::Read<Depot>>().unwrap();
+    let lock = req.get::<persistent::State<Depot>>().unwrap();
+    let mut depot = lock.write().expect("depot write lock is poisoned");
     let session = req.extensions.get::<Authenticated>().unwrap();
     let (channel, ident) = {
         let params = req.extensions.get::<Router>().unwrap();
-        let channel = params.find("channel").unwrap();
-        let ident = ident_from_params(params);
+        let origin = match params.find("origin") {
+            Some(o) => o.to_string(),
+            _ => return Ok(Response::with(status::BadRequest)),
+        };
+
+        let mut conn = Broker::connect().unwrap();
+
+        if try!(get_origin(&mut conn, &origin)).is_none() {
```

Comment:
I think we do check for origin access in this function, down on line 1130, right?  I'll add that to create/delete channel though.